core/src/main/scala/chisel3/internal/MonoConnect.scala

```diff
@@ -178,6 +188,35 @@ private[chisel3] object MonoConnect {
       case (sink, source) => throw MismatchedException(sink.toString, source.toString)
     }
 
+  /** Check whether two [[Vec]]s can be bulk connected (<-) in FIRRTL. */
+  private[chisel3] def canBulkConnectVecs(sink_v: Vec[Data @unchecked],
+                                          source_v: Vec[Data @unchecked],
+                                          sourceInfo: SourceInfo): Boolean = {
+    val canWriteBinding = source_v.topBinding match {
+      case _: ReadOnlyBinding => false
+      case _ => true
+    }
+    val elemsMatch = CheckTypes.validConnect(
+      Converter.extractType(sink_v, sourceInfo),
```

Comment:
do the Vecs need to be parameterized exactly in order to be bulk connected? is that what is captured in `extractType`?

Reply:
They need not be exact, implicit truncation remains the law of the land and that is correctly handled by `validConnect`.

@debs-sifive One thing that we also need to check is the Flows. The method needed for this is not exposed from FIRRTL so we might need to make some utility methods in FIRRTL: https://github.com/chipsalliance/firrtl/blob/4991ee209b4ae179afcb3e043bbe7ba92006931d/src/main/scala/firrtl/passes/CheckFlows.scala#L64. This is a bit tricky because those seem to require Kind and throw exceptions so we can't just reuse out of the box anyway

For development purposes, we can just write these utilities in Chisel and then migrate them to FIRRTL.

We also need to check "visibility" (eg. https://github.com/chipsalliance/chisel3/blob/b6929c9ad438db26055707b8ec5c66e4f70a22b8/core/src/main/scala/chisel3/internal/MonoConnect.scala#L217, need to make sure we're not issuing connects if one of the sides is not part of the current module).